airbyte-integrations/connectors/source-microsoft-sharepoint/source_microsoft_sharepoint/utils.py

```diff
@@ -0,0 +1,75 @@
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+
+import time
+from datetime import datetime, timedelta
+from http import HTTPStatus
+
+from airbyte_cdk.models import FailureType
+from airbyte_cdk.sources.file_based.remote_file import RemoteFile
+from airbyte_cdk.utils.traced_exception import AirbyteTracedException
+
+
+class MicrosoftSharePointRemoteFile(RemoteFile):
+    download_url: str
+
+
+def filter_http_urls(files, logger):
+    for file in files:
+        if file.download_url.startswith("http") and not file.download_url.startswith("https"):  # ignore-https-check
+            logger.error(f"Cannot open file {file.uri}. The URL returned by SharePoint is not secure.")
+        else:
+            yield file
+
+
+def execute_query_with_retry(obj, max_retries=5, initial_retry_after=5, max_retry_after=300, max_total_wait_time=600):
+    """
+    Executes a query with retry logic on encountering specific HTTP errors.
+
+    This function attempts to execute `obj.execute_query()` method, applying exponential backoff
+    retry logic for HTTP status codes 429 (Too Many Requests) and 503 (Service Unavailable). It
+    respects the 'Retry-After' header from the response, if present.
+
+    Parameters:
+        obj (object): The object that has the `execute_query` method to be executed.
+        max_retries (int): Maximum number of retry attempts. Defaults to 5.
+        initial_retry_after (int): Initial waiting time (in seconds) before the first retry. Defaults to 5 seconds.
+        max_retry_after (int): Maximum waiting time (in seconds) between retries. Defaults to 300 seconds.
+        max_total_wait_time (int): Maximum total waiting time (in seconds) for all retries. Defaults to 600 seconds.
+
+    Raises:
+        AirbyteTracedException: If the maximum total wait time or the maximum number of retries is exceeded.
+
+    Returns:
+        The result of `obj.execute_query()` if successful within the retry constraints.
+    """
+    retries = 0
+    start_time = datetime.now()
+    retry_after = initial_retry_after
+
+    while retries < max_retries:
+        try:
+            return obj.execute_query()
+        except Exception as ex:
+            if hasattr(ex, "response") and ex.response.status_code in (HTTPStatus.TOO_MANY_REQUESTS, HTTPStatus.SERVICE_UNAVAILABLE):
+                current_time = datetime.now()
+                elapsed_time = (current_time - start_time).total_seconds()
+
+                retry_after_header = ex.response.headers.get("Retry-After", None)
+                if retry_after_header:
+                    retry_after = int(retry_after_header)
+
+                if elapsed_time + retry_after > max_total_wait_time:
+                    message = f"Maximum total wait time of {max_total_wait_time} seconds exceeded for execute_query."
```

Comment:
As a developer troubleshooting an error, how do I know what is the underlying issue? Should we add the reason why it failed and not just the number of retries? Assuming we are handling `ClientRequestException` here, it seems like there is a `message` attribute that could be relevant

Reply:
Fixed